### ISSUE TITLE
Region Failover logic improvements

### ISSUE THREAD

**Issue by j82w** [Azure/azure-cosmos-dotnet-v3#2489]
This issue is to discuss/design Partition level failover for read heavy or read-only workloads.

The current design for write operations is it will mark a single partition as unavailable. 
- If it returns an exception that guarantees the write did not occur like a write forbidden the SDK will retry the request on the secondary regions until it finds the region where the request succeeds which means it is the current primary region.
- If it returns an exception that it is ambiguous like 500 internal server error or a 408 request timeout then it marks the partition unavailable and routes future requests to the next region. The current exception is returned to user since it is not safe to retry.

The problem:
On write requests there is a very clear write forbidden exception that is thrown if it is not the current primary. This allows the SDK to round robin to other regions to find the new primary region for that partition since all the other regions will return write forbidden exception.
Read requests are allowed on all regions. If the primary region is down it will route all requests to the secondary region. The problem is that there is no clear signal that the primary region is now back to a healthy state and to switch back to the primary region.

The current SDK region failover logic marks a region unavailable and will go to a secondary region for 5 minutes. After the 5 minutes it will remove the flag and send request back to the primary region. If it is still down the requests will fail again and the region will be marked as unavailable again for another 5 minutes.

This is a very long time to route request to a secondary region if it is not necessary. This also causes requests at the 5 minutes mark to have an increase latency and the possibility of failing if the region is not backup.

Possible solutions:
1. Have the Cosmos DB service return a new header which is a flag if it is the primary region for that partition. The SDK can look at the response header to see when the partition failed back over to the other region. It can use the same round robin logic based on the response header as the write forbidden to find the new primary region. It can further be expanded to possible include a hit on the correct primary region, but it should only be considered a hint since it may be out of date.
   Pros: Fast to fail back over, follows the same failover logic as writes, no impact to user requests once it fails over
   Cons: Requires backend changes.
2. Start a background thread which does a point read to a non-existing document that checks if the primary region is backup for that partition. The background thread would run every 30 seconds(exact interval needs to be discussed). Once the region is backup it would mark the region as available and all user requests would again be routed to the primary region.
    Pros: Fast to fail back over and no user impact from trying the primary region
    Cons: Requires a background thread and may result in additional RU charge
3. Randomly send a user request every so often to the primary region. This requires the user to send a point read operation so it is fast and retriable. 
    Pros: Fast to fail back over 
    Cons: Will impact request latency of selected requests that try to go to primary region and requires user to send requests. Sparse workloads maybe impacted more from this.

**Comment by j82w** [comment]
Non-distinguished (example: read operation could be completed on any region)
Distinguished (example: write for single master must go to specific region)

Retriable operations (example: read operations)
Non-Retriable operation (example: write operations)

Retriable exception (example: 410 Gone)
Non-Retriable exception (example: 408 request timeout)

**Comment by j82w** [comment]
Rough draft on different scenarios and how they will be handled.
All scenarios assume the account has multiple regions configured.

Distinguished 
Non-Retriable operation
Non-Retriable exception:
(Timeout exception on a write operation for a single master account)
--> Write request timeout against R1
--> R1 is marked unavailable for writes. Future write request go to R2.
--> R2 request succeeds. 
----> Nothing changes. Request continue to be routed to R2.
--> R2 request fails. Round robin to other secondary regions until success.
----> R2 is marked as unavailable for writes. Target the next region based on priority list.

Non-distinguished 
Non-Retriable operation
Non-Retriable exception:
(Timeout exception on a write operation for a multi-master account)
--> Write request timeout against R1
--> R1 is marked unavailable for writes. Future write request go to R2.
--> R2 request succeeds. 
----> Request continue to be routed to R2.
----> Background thread is started to do a head request against primary replica every 10 seconds.
-------> On success a random write operation will be chosen and sent to the primary region to validate the health. If success the region will be marked available. If failure the background job will sleep and try again in the future.
--> R2 request fails. Round robin to other secondary regions until success.
----> R2 is marked as unavailable for writes. Target the next region based on priority list.

Non-distinguished 
Non-Retriable operation
Retriable exception:
(410 on a write operation for a multi-master account)
--> Write request timeout against R1
--> Request is retried on R1 after refreshing caches

Non-distinguished 
Retriable operation
Retriable exception:
(Timeout exception on a read operation)
--> Read request timeout against R1
--> R1 is marked unavailable for reads and writes (if read fails unlikely write will succeed so mark both unavailable to reduce failures on non-retriable write operations). Future requests go to R2.
--> R2 request succeeds. 
----> Request continue to be routed to R2.
----> Background thread is started to do a head request against primary replica every 10 seconds.
-------> On success a random read operation will be chosen and sent to the primary region to validate the health. If success the region will be marked available for reads. If failure the background job will sleep and try again in the future.
--> R2 request fails. Round robin to other secondary regions until success.
----> R2 is marked as unavailable. Target the next region based on priority list.

**Comment by kirankumarkolli** [comment]
@J82w what's are the scoped next steps we need to proceed furhter?

**Comment by NaluTripician** [comment]
Superseded by PPAF/PPCB
